`[Bugfix | Issue #2565]` Network Switching and Context to Store update

Da-Colon opened the pull request:
Closes #2565

## Updates the Network Provider to zustand store
Now switching networks don't cause a massive tree re-render

## Fixes App Crash when navigating to a DAO on a different network via links, refresh or external link
We will need to test with other wallets, but now that 'being connected to a network' is decoupled from the choice in the wallet we don't need to use `switchchain` anymore

Da-Colon commented on src/providers/NetworkConfig/NetworkConfigProvider.tsx:
Deleted...

Da-Colon commented on src/providers/NetworkConfig/useNetworkConfigStore.ts:
```diff
@@ -0,0 +1,31 @@
+
+// Create the Zustand store
+export const useNetworkConfigStore = create<NetworkConfigStore>(set => ({
+  ...sepoliaConfig,
```
defaults to sepolia...though probably should set this up pending on env? the only place this would be noticeable would be the network switcher...

Speaking of....we can probably get rid of that.....

and if we get rid of that we should probably add something somewhere to indicate what network you are on more explictly

Da-Colon replied:
@decentdao/design

adamgall reviewed (commented):
Maybe it's just me, but selecting a network in the in-app network switcher doesn't work. I start on Sepolia, try to switch to Base, but when I open the dropdown again I'm still on Sepolia.

Da-Colon commented:
I didn't update the network switcher to work with this setup. The app no longer cares what network the wallet is 'connected' to

Da-Colon commented:
Discussed further with @nicolaus-sherrill the current plan is:

- Rip out Network Selector from Wallet Menu
- Add Network Selector to `DAO Essentials`. 
- Add Network indictor (icon?) to DAO Home

nicolaus-sherrill commented:
@Da-Colon 

- Network switcher removed from top nav components.
<img width="772" alt="Screenshot 2024-12-05 at 4 26 33 PM" src="https://github.com/user-attachments/assets/a70e73e3-3e34-4be3-9617-e2690a058fde">

- Network icon added to DAO dashboard info tile. Component updated with padding reduced to 1rem and favorite icon shifted to right, next to settings. [Component linked here](https://www.figma.com/design/ur5hElxplkmlY9gnGKlwrm/Components-%26-Styles---Decent-Design-System-1.1.1?m=auto&node-id=2248-25539&t=u0dzpMgPBzXdWZzc-1)

<img width="887" alt="Screenshot 2024-12-05 at 4 30 53 PM" src="https://github.com/user-attachments/assets/ac5e4683-293b-4f5e-8a0d-765423df915d">

- Create/edit a DAO workflow updated to include the network selector at the beginning of the flow. [Ready in Handoff.](https://www.figma.com/design/9hUYtvCKjOyB92vt8ukYtD/Product-Handoff?m=auto&node-id=7208-30777&t=rTmswoFYVnJHrYhy-1)
- Selector dropdown pill component updated with a token and a network variant [here](https://www.figma.com/design/ur5hElxplkmlY9gnGKlwrm/Components-%26-Styles---Decent-Design-System-1.1.1?m=auto&node-id=4329-422&t=u0dzpMgPBzXdWZzc-1).
- - (this will be consolidated with other component classes in the near future)

Da-Colon commented:
> @Da-Colon
> 
> * Network switcher removed from top nav components.
> 
> <img alt="Screenshot 2024-12-05 at 4 26 33 PM" width="772" src="https://private-user-images.githubusercontent.com/62394931/393019720-a70e73e3-3e34-4be3-9617-e2690a058fde.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM0Mzg3NTgsIm5iZiI6MTczMzQzODQ1OCwicGF0aCI6Ii82MjM5NDkzMS8zOTMwMTk3MjAtYTcwZTczZTMtM2UzNC00YmUzLTk2MTctZTI2OTBhMDU4ZmRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA1VDIyNDA1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiZjhiODY0NTQ0NjBjZWE3MDFhMmExZjJjM2ZkNmY0ZjBlNDViMzg2MzhkZmVjYTYzMTUwODUzMzkwNzYzOTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6zufQ-q-IopeyqH21s4N3ETneuiKj8YzwknQ00Hr1FY">
> * Network icon added to DAO dashboard info tile. Component updated with padding reduced to 1rem and favorite icon shifted to right, next to settings. [Component linked here](https://www.figma.com/design/ur5hElxplkmlY9gnGKlwrm/Components-%26-Styles---Decent-Design-System-1.1.1?m=auto&node-id=2248-25539&t=u0dzpMgPBzXdWZzc-1)
> 
> <img alt="Screenshot 2024-12-05 at 4 30 53 PM" width="887" src="https://private-user-images.githubusercontent.com/62394931/393020666-ac5e4683-293b-4f5e-8a0d-765423df915d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM0Mzg3NTgsIm5iZiI6MTczMzQzODQ1OCwicGF0aCI6Ii82MjM5NDkzMS8zOTMwMjA2NjYtYWM1ZTQ2ODMtMjkzYi00ZjVlLThhMGQtNzY1NDIzZGY5MTVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA1VDIyNDA1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBjY2E4OGNiZjg0Njc3N2MxM2FlNmY0NTQyNjIzNjEzNzM3MGY1ZTQwMmJmOGI5ZTY2OGU2MzhlMmM5NmIxZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DmMtr8FCxYGYpi-pUBkcju0yEo1ip3XQbEKR69dlOEE">
> * Create/edit a DAO workflow updated to include the network selector at the beginning of the flow. [Ready in Handoff.](https://www.figma.com/design/9hUYtvCKjOyB92vt8ukYtD/Product-Handoff?m=auto&node-id=7208-30777&t=rTmswoFYVnJHrYhy-1)
> * Selector dropdown pill component updated with a token and a network variant [here](https://www.figma.com/design/ur5hElxplkmlY9gnGKlwrm/Components-%26-Styles---Decent-Design-System-1.1.1?m=auto&node-id=4329-422&t=u0dzpMgPBzXdWZzc-1).
> * * (this will be consolidated with other component classes in the near future)

fantastic, thank you sir

adamgall commented:
How do I search for a DAO on a specific network?

Like, when I paste the Base Myosin DAO address into the search bar, it tells me "this dao doesn't exist on Ethereum". How do I tell it to search on Base?

adamgall commented on src/components/Roles/forms/RoleFormAssetSelector.tsx:
The changes in this file seem unrelated to this PR. Are these intentional?

Da-Colon replied:
Yes, There were UI updates to the DropdownSelector. I can if needed branch out the last couple of commits which:
- Extract the UI from the Asset Dropdown into a reusable component
- Update the UI for it
- reuse for Network Selector in Create DAO workflow
- Add Network icon to DAO Dashboard

adamgall replied:
Ok, as long as these were intentional changes, all good. Thanks.

adamgall commented:
When attempting to deploy a new DAO, I selected "Sepolia" as the network to deploy on.

However, when getting to the end of the form and attempting to deploy, my wallet (Frame) popped up with a Base transaction for me to sign.

adamgall commented:
What is the justification for this PR?

I'm not familiar with any dapps that _don't_ have an in-app network switcher. Can you point to an example or two of "multi-chain dapps" that follow this same pattern of _not_ including in-app network switchers?

Da-Colon commented:
> How do I search for a DAO on a specific network?
> 
> Like, when I paste the Base Myosin DAO address (0x374D5a17D2433415C48C735A493F0523DaB99041) into the search bar, it tells me "Sorry, this address is not a DAO on Ethereum.". How do I tell it to search on Base?



> What is the justification for this PR?
> 
> I'm not familiar with any dapps that _don't_ have an in-app network switcher. Can you point to an example or two of "multi-chain dapps" that follow this same pattern of _not_ including in-app network switchers?

So working through some solutions to solve this without being to intrusive. basically `wagmi` is still looking at its config for what network to use for its hooks `usePublicClient`, `useWalletClient`, etc..

Da-Colon commented:
> How do I search for a DAO on a specific network?
> 
> Like, when I paste the Base Myosin DAO address (0x374D5a17D2433415C48C735A493F0523DaB99041) into the search bar, it tells me "Sorry, this address is not a DAO on Ethereum.". How do I tell it to search on Base?

crap good call. I'll need to add a simple network switcher for the DAO search too

Da-Colon commented:
So I can't point to any patterns of this out there. But I'm not one to stop at that. 

So justification started at fixing the cross network communication in the app. and I do apologize as the PR contained a little past just fixing that first issue.

So first thing I did was what i described in the description moving from context to store. This did not fix the `switchchain` issue in the app. I explore a few other adjusts causing the app to crash in 3 scenarios. with adjustments here and there I was able to solve 1 or 2. but never all 3. and code started looking funky.

After some thought. I realize that we don't need to rely on the wallet switching chain to read data on a different network just switch up the config.

We will still need to 'switchchain' but this can now target when a user goes to do an action. if they are viewing a DAO and their wallet isn't connected to the right network they will get the pop. this way we don't have to dynamically change the network wallet everytime a user is visitng the site or just viewing a DAO.

adamgall commented:
> > How do I search for a DAO on a specific network?
> > Like, when I paste the Base Myosin DAO address (0x374D5a17D2433415C48C735A493F0523DaB99041) into the search bar, it tells me "Sorry, this address is not a DAO on Ethereum.". How do I tell it to search on Base?
> 
> crap good call. I'll need to add a simple network switcher for the DAO search too

Maybe a better pattern would be to search all networks at the same time, and show all matches in the results dropdown if there are more than one.

Da-Colon commented:
> > > How do I search for a DAO on a specific network?
> > > Like, when I paste the Base Myosin DAO address (0x374D5a17D2433415C48C735A493F0523DaB99041) into the search bar, it tells me "Sorry, this address is not a DAO on Ethereum.". How do I tell it to search on Base?
> > 
> > 
> > crap good call. I'll need to add a simple network switcher for the DAO search too
> 
> Maybe a better pattern would be to search all networks at the same time, and show all matches in the results dropdown if there are more than one.

Honestly, yeah you are right we sure can do that

Da-Colon commented:
@adamgall still working though how to get warming to listen to the app rather than wallet with having to have a pass through hook that add chainID. Gonna break and eat something

Da-Colon commented:
Summary from the standup call: 

- Go ahead with UI/Design updates
- This PR is a little got a little bloated but contains:
  - Moves NetworkConfig Context Provider -> Zustand store
  - Updates all components imports to store
  - Removes Network Dropdown from wallet menu
  - Adds Network Icon to DAO Home
  - Removes `useSwitchChain` from Internal DAO navigation components (MySafes, SafeList)
  - Adds UI for Network Switching in DAO Creation
  
  I've branched off for final updates to test and make sure everything is working see #2609 For this where I'll be adding any work to make sure this all works including:
  - Making sure that the underlying network used by `publicClient` is properly switched.
  - testing different wallets to ensure that network is changed correctly.

parkermccurley commented:
This is such a valuable effort, thank you @Da-Colon!

DarksightKellar commented on src/pages/dao/SafeController.tsx:
```diff
@@ -20,17 +20,18 @@ import LoadingProblem from '../LoadingProblem';
   const { errorLoading } = useFractalNode({
     addressPrefix,
     safeAddress,
+    wrongNetwork,
```
Instead of passing this in here, can `const { wrongNetwork } = useParseSafeAddress()` not be directly used in `useFractalNode`?

Da-Colon replied:
I personally prefer that the value being used when the app makes this decision, versus random places in the app that are doing the same parsing

DarksightKellar replied:
> I personally prefer that the value being used when the app makes this decision, versus random places in the app that are doing the same parsing

wait say that again? Not sure what you're saying there

Da-Colon replied:
lol sorry, just prefer to pass as a param versus using `useSearchParams` and `useParseAddress`  imports inside of useFractalNode when it can just as easily be the same value that is used to throw errors.

DarksightKellar replied:
Ah, fair

Da-Colon commented on src/components/ui/menus/DropdownMenu.tsx:
```diff
@@ -0,0 +1,306 @@
+  renderItem?: (item: DropdownItem<T>, isSelected: boolean) => React.ReactNode;
+}
+
+export function DropdownMenu<T>({
```
That'd be the goal but not for this PR. And we need to make sure we have a clear idea of any 'variants' that are needed by cross referencing design. I don't want to start any trend of retro fitting things to work in other places

Da-Colon replied:
https://github.com/orgs/decentdao/projects/32/views/3?pane=issue&itemId=90416867

DarksightKellar replied:
Ah gotcha

Da-Colon commented on src/components/ui/menus/DropdownMenu.tsx:
```diff
@@ -0,0 +1,306 @@
+  renderItem?: (item: DropdownItem<T>, isSelected: boolean) => React.ReactNode;
+}
+
+export function DropdownMenu<T>({
```
eh. Why brand it?

DarksightKellar replied:
Not brand per se, just want a name that doesn't read like something imported from a package or sth

Da-Colon replied:
Let's go at this when we tackle #2620

adamgall reviewed (changes requested):
Have an issue loading a DAO name on Base, in the search menu.

Here's how I got to this situation:
- fresh instance of https://deploy-preview-2595.app.dev.decentdao.org/
- searched for "decentdao.eth" -> it popped up and the name loaded -> clicked it to go to the DAO. Saved it as a "Favorite".
- went back to home page
- pasted in `0x374D5a17D2433415C48C735A493F0523DaB99041` to the search bar, which is the Base Myosin DAO address
- screenshot 👇 

![Screenshot 2024-12-16 at 2 24 36 PM](https://github.com/user-attachments/assets/88d3e997-a025-4e3a-95d5-4ec90a8f24bb)

Da-Colon commented:
> * https://deploy-preview-2595.app.dev.decentdao.org/

Thanks on it.

Da-Colon commented:
weird locally its working fine and can't reproduce. @adamgall

Da-Colon commented:
@adamgall wha? on the preview 

![image](https://github.com/user-attachments/assets/cc123073-eafd-4826-bc1c-bbed807ceab3)

then why let me do it locally? you don't need a play by play ha.

DarksightKellar reviewed (approved):
Approving, but something's off with Base Myosin DAO. Navigating directly to `https://deploy-preview-2595.app.dev.decentdao.org/home?dao=base:0x374D5a17D2433415C48C735A493F0523DaB99041` results in a white screen of death. Similar weird behaviour on dev and prod actually. However clicking into the DAO from saved list (after searching if on fresh cache) works fine.

adamgall commented:
> https://deploy-preview-2595.app.dev.decentdao.org/home?dao=base:0x374D5a17D2433415C48C735A493F0523DaB99041

I'm not getting any issues with this direct link... @DarksightKellar can you take a video of what you're experiencing

Da-Colon commented:
@DarksightKellar Thanks, I can reproduce. On it

DarksightKellar commented:
> @DarksightKellar Thanks, I can reproduce. On it

Weird lol. Curious what's going on there

Da-Colon commented:
@DarksightKellar I figured it out. Soon I saw the white screen I suspected something was up with around the wallet and it checking the connection. if it tries to switch chain prior to being ready, I believe it catches it self in a loop. I think the first attempt (first commit I just did) adding `isFetched` worked and I got eager and tested before it was deployed, I switched to `isFetchedAfterMount` which makes sense anyway here.

Anyway, fixed

DarksightKellar commented:
Works!

@Da-Colon  In other news, would you know anything about this intermittent toast? Also only on myosin:

<img width="596" alt="Screenshot 2024-12-17 at 18 36 54" src="https://github.com/user-attachments/assets/d402341c-da6a-48c9-b5a3-d02e418ccf69" />

Da-Colon commented:
yeah preexisting. I had made some updates, but I guess there is still something trying to load in state and not getting reset. I'll take a look
